LibOS/shim/test/regression/epoll_wait_timeout.c

```diff
@@ -0,0 +1,119 @@
+/* -*- mode:c; c-file-style:"k&r"; c-basic-offset: 4; tab-width:4; indent-tabs-mode:nil; mode:auto-fill; fill-column:78; -*- */
+/* vim: set ts=4 sw=4 et tw=78 fo=cqt wm=0: */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/epoll.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netdb.h>
+
+static int create_and_bind (char *port) {
+    struct addrinfo hints;
+    struct addrinfo *result, *rp;
+    int s, sfd;
+
+    memset (&hints, 0, sizeof (struct addrinfo));
+    hints.ai_family = AF_UNSPEC;     /* Return IPv4 and IPv6 choices */
+    hints.ai_socktype = SOCK_STREAM; /* We want a TCP socket */
+    hints.ai_flags = AI_PASSIVE;     /* All interfaces */
+
+    s = getaddrinfo (NULL, port, &hints, &result);
+    if (s != 0) {
+        fprintf (stderr, "getaddrinfo: %s\n", gai_strerror (s));
+        return -1;
+    }
+
+    for (rp = result; rp != NULL; rp = rp->ai_next) {
+        sfd = socket (rp->ai_family, rp->ai_socktype, rp->ai_protocol);
+        if (sfd == -1)
+            continue;
+
+        s = bind (sfd, rp->ai_addr, rp->ai_addrlen);
+        if (s == 0) {
+            /* We managed to bind successfully! */
+            break;
+        }
+
+        close (sfd);
+    }
+
+    if (rp == NULL) {
+        fprintf (stderr, "Could not bind\n");
+        return -1;
+    }
+
+    freeaddrinfo (result);
+
+    return sfd;
+}
+
+static int make_socket_non_blocking (int sfd) {
+    int flags, s;
+
+    flags = fcntl (sfd, F_GETFL, 0);
+    if (flags == -1) {
+        perror ("fcntl");
+        return -1;
+    }
+
+    flags |= O_NONBLOCK;
+    s = fcntl (sfd, F_SETFL, flags);
+    if (s == -1) {
+        perror ("fcntl");
+        return -1;
+    }
+
+    return 0;
+}
+
+#define MAXEVENTS 64
+
+int main (int argc, char *argv[]) {
+    int sfd, s, n;
+    int efd;
+    struct epoll_event event;
+    struct epoll_event *events;
+
+    sfd = create_and_bind("8000");
```

Comment:
You may need to make the port a command-line argument.  We have had a few issues with jenkins where, if the same ephemeral port is used in quick succession, the test will spuriously fail.